Ukol_2/cpp/sifry.cpp

```diff
@@ -1,72 +1,139 @@
-{
-  // Implementace Vigenerovy šifry
-  // sifrovat = true pro šifrování, sifrovat = false pro dešifrování
-  return "";
+// Funkce pro Caesarovu šifru
+std::string caesar_sifra(const std::string &text, int posun, bool sifrovat) {
+    std::string vysledek;
+    int smer = sifrovat ? 1 : -1;
+
+    for (char znak : text) {
+        if (std::isalpha(znak)) {
+            char base = std::islower(znak) ? 'a' : 'A';
+            vysledek += (znak - base + smer * posun + 26) % 26 + base;
+        } else {
+            vysledek += znak;
+        }
+    }
+
+    return vysledek;
 }
 
-// Funkce pro XOR šifru
-std::string xor_sifra(const std::string &text, const std::string &klic, bool sifrovat)
-{
-  // Implementace XOR šifry
-  // sifrovat = true pro šifrování, sifrovat = false pro dešifrování
-  return "";
+// Funkce pro Vigenerovu šifru (šifrování i dešifrování)
+std::string vigener_sifra(const std::string &text, const std::string &klic, bool sifrovat) {
+    std::string vysledek;
+    size_t delka_klice = klic.size();
+    int smer = sifrovat ? 1 : -1;
+
+    for (size_t i = 0; i < text.size(); ++i) {
+        if (std::isalpha(text[i])) {
+            char base = std::islower(text[i]) ? 'a' : 'A';
+            char klicove_znak = klic[i % delka_klice];
+            char upraveny_klíč = std::islower(klicove_znak) ? klicove_znak - 'a' : klicove_znak - 'A';
+            vysledek += (text[i] - base + smer * upraveny_klíč + 26) % 26 + base;
+        } else {
+            vysledek += text[i];
+        }
+    }
+
+    return vysledek;
 }
 
-// Funkce pro uložení řetězce do souboru
-void uloz_do_souboru(const std::string &jmeno_souboru, const std::string &obsah)
-{
-  // Implementace funkce pro uložení řetězce do souboru
+// Funkce pro XOR šifru (šifrování i dešifrování)
+std::string xor_sifra(const std::string &vstup, const std::string &klic, bool sifrovat) {
+    if (klic.empty()) {
+        throw std::invalid_argument("Heslo nesmí být prázdné!");
+    }
+
+    std::string vysledek;
+    size_t delka_klice = klic.size();
+
+    if (sifrovat) {
+        // Šifrování - převod na binární formát
+        for (size_t i = 0; i < vstup.size(); ++i) {
+            char sifrovany_znak = vstup[i] ^ klic[i % delka_klice];
+            for (int j = 7; j >= 0; --j) {
+                vysledek += ((sifrovany_znak >> j) & 1) ? '1' : '0';
+            }
+        }
+    } else {
+        // Dešifrování - binární formát na původní text
+        if (vstup.size() % 8 != 0) {
+            throw std::invalid_argument("Binární text musí být násobkem 8!");
+        }
+
+        for (size_t i = 0; i < vstup.size(); i += 8) {
+            char binarni_znak = 0;
+            for (int j = 0; j < 8; ++j) {
+                binarni_znak = (binarni_znak << 1) | (vstup[i + j] - '0');
+            }
+            char puvodni_znak = binarni_znak ^ klic[(i / 8) % delka_klice];
+            vysledek += puvodni_znak;
+        }
+    }
```

Comment:
`for (size_t i = 0; i < vstup.size(); ++i){
vysledek += vstup[i] ^ klic[i % delka_klice];
}`